src/index.ts

```diff
@@ -83,77 +75,114 @@ export function latexRequest<T>(
+    const baseName = PathExt.basename(texContext.path, '.tex');
+    const pdfFilePath = PathExt.join(dirName, baseName + '.pdf');
+
+    let pdfContext: DocumentRegistry.IContext<DocumentRegistry.IModel>;
+    let errorPanel: ErrorPanel | null = null;
+
+    // Hook up an event listener for when the '.tex' file is saved.
+    texContext.fileChanged.connect((sender, args) => {
+      latexRequest(texContext.path, serverSettings).then(() => {
+        // Read the pdf file contents from disk.
+        if (pdfContext) {
+          pdfContext.revert();
+        } else {
+          const pdfWidget = manager.openOrReveal(pdfFilePath);
+          pdfContext = manager.contextForWidget(pdfWidget);
+        }
+        if (errorPanel) {
+          errorPanel.close();
+        }
       }).catch((err) => {
         // If there was an error, read the log
         // file from disk and show it.
-        errorPanel = new ErrorPanel();
+        if (!errorPanel) {
+          errorPanel = Private.createErrorPanel();
+          // On disposal, set the reference to null
+          errorPanel.disposed.connect(() => {
+            errorPanel = null;
+          });
+          //Add the error panel to the main area.
+          app.shell.addToMainArea(errorPanel, { ref: widget.id });
+        }
         errorPanel.text = err.xhr.response;
-        errorPanel.id = `latex-error-${++Private.id}`;
-        errorPanel.title.label = 'LaTeX Error';
-        errorPanel.title.closable = true;
-        app.shell.addToMainArea(errorPanel, { ref: widget.id });
       });
+    });
+
+    // Run an initial latexRequest so that the appropriate files exist,
+    // then open them.
+    latexRequest(texContext.path, serverSettings).then(() => {
+      // Open the pdf and get a handle on its document context.
+      const pdfWidget = manager.openOrReveal(pdfFilePath);
+      pdfContext = manager.contextForWidget(pdfWidget);
+    }).catch((err) => {
+      // If there was an error, read the log
+      // file from disk and show it.
+      errorPanel = Private.createErrorPanel(err.xhr.response);
+      // On disposal, set the reference to null
+      errorPanel.disposed.connect( () => {
+        errorPanel = null;
+      });
+      //Add the error panel to the main area.
+      app.shell.addToMainArea(errorPanel, { ref: widget.id });
+    });
+
+    // When the tex file is closed, remove it from the cache.
+    // The listener should be removed in its own dispose() method.
+    texContext.disposed.connect(() => {
+      Private.previews.delete(texContext.path);
+    });
+
+    // Update the set of active previews and cache the values.
+    Private.previews.add(texContext.path);
+    state.save(id, { paths: Array.from(Private.previews) });
+  };
+
+  // If there are any active previews in the statedb,
+  // activate them upon initialization.
+  Promise.all([state.fetch(id), app.restored]).then(([args]) => {
+    let paths = (args && args['paths'] as string[]) || [];
```

Comment:
what does as string[] do?

Reply:
This is a place where the typescript compiler can't infer the type. I am letting it know it is a list of strings.